Placeholder user services and administration

wielinde opened the pull request:
Bases on PR #8935.

- [x] Adds basic contracts and services for working with placeholder users.
- [x] Allows basic administration of placeholder users
  - [x] Query placeholder users
  - [x] Placeholder user: Create, Update
  - [x] Memberships: Create, Update, Delete
    
Out of scope:
- Deletion of placeholder users (should get done in a separate PR)
- Assign placeholder users to groups

wielinde committed:
Adding placeholder user contracts

wielinde committed:
Adding create, update, and delete services for placeholder users

ulferts commented on app/contracts/placeholder_users/delete_contract.rb:
```diff
@@ -0,0 +1,60 @@
+require 'placeholder_users/base_contract'
+
+module PlaceholderUser
+  class DeleteContract < BaseContract
```
This should use the BaseServices::DeleteContract superclass (see https://github.com/opf/openproject/pull/8937/commits/9ef1c42c437bbfb48cdbc755599399c1abdc8e92)

ulferts replied:
Still true

wielinde replied:
outsch.

wielinde committed:
WIP: Adding Placeholder User contract specs [ci skip]

oliverguenther committed:
Extract contract validation into common helper

oliverguenther committed:
Add common validation in BaseContract + common example for admin checks

oliverguenther committed:
Introduce common ModelContract shared context for validations

wielinde committed:
WIP: PlaceholderUser controller, i18n, and routes [ci skip]

wielinde committed:
Placeholder users index page and query

- moved all group related scopes from User to Principal to make them also available in PlaceholderUser.

wielinde committed:
end

wielinde committed:
Create PlaceholderUser

wielinde committed:
Feature spec for editing a placeholder user

wielinde committed:
Manage PlaceholderUser memberships

The managment of memberships is pretty similar for User and PlaceholderUser. This commit extacts the similarities and uses them for both.

wielinde committed:
General partial and show view for PlaceholderUser

wielinde committed:
Merge remote-tracking branch 'origin/dev' into feature/placeholder-user-services

wielinde committed:
Delete obosolete partial

oliverguenther committed:
Allow RequireAdminGuard to be used as a module function

wielinde committed:
Fix I18n for confirmation text

wielinde committed:
Smaller code improvements

wielinde committed:
Fix: Syntax for accessing status enums was wrong.

wielinde committed:
Use UpdateService for updating a placeholder user

ulferts commented on app/cells/views/individual_principal_base_filter/show.erb:
```diff
@@ -37,11 +37,8 @@ See docs/COPYRIGHT.rdoc for more details.
-      </li>
-      <% if groups.present? %>
+      <% if groups.present? &&
+            self.class != PlaceholderUsers::PlaceholderUserFilterCell %>
```
Why is this done? It breaks inheritance.

wielinde replied:
That is a temporary hack that we both agreed on to do until placeholder users can be added to groups.

ulferts replied:
Why not causing `groups` to be nil for the time being?

wielinde replied:
addressed here https://github.com/opf/openproject/pull/8944/commits/29aa048008b27aa5e6d05a35efb531586a4892be

ulferts commented on app/contracts/placeholder_users/base_contract.rb:
```diff
@@ -0,0 +1,57 @@
+
+module PlaceholderUsers
+  class BaseContract < ::ModelContract
+    attribute :type
```
This should not be exposed. In effect, it will allow altering the type of the user, e.g. turning it into a real User, whenever somebody finds a way to sneak in a value for type through the controllers/api endpoints. 

Instead, the type should only be set via the system and should be marked as such, please see https://github.com/opf/openproject/blob/dev/app/services/messages/set_attributes_service.rb#L43 for an example

ulferts replied:
A different approach could be to move the declaration for `type` into the `CreateContract`. That way, at least existing `PlaceholderUser`s cannot be altered. And then add a validation on top of it to ensure that the value can only be `"PlaceholderUser"`

wielinde replied:
Thank you for the clarification. I was not aware of all that when writing/copying. I will remove it here. However, could it be that we also have an issue of type exposure here? https://github.com/opf/openproject/blob/3c9be3bdbe7dce96b070491a6a4b8de34fd5f7b0/app/contracts/users/base_contract.rb#L35

ulferts replied:
Yes, good catch. This should also be reworked, then.

wielinde replied:
I removed it from the Users::BaseContract

wielinde replied:
Ok, had to tackle it again as validations did not pass anymore. This whole thing smells a bit as I understand that as a single table inheritance topic that should not be exposed to higher layers.

My current solution is to only allow writing of `:type` if `model.new_record?` is true.

ulferts commented on app/contracts/placeholder_users/create_contract.rb:
```diff
@@ -0,0 +1,36 @@
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+require 'placeholder_users/base_contract'
```
This should be superfluous.

wielinde replied:
True. Was copy and paste from the user contracts. I fixed it there, too.

ulferts commented on app/contracts/placeholder_users/base_contract.rb:
```diff
@@ -0,0 +1,57 @@
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+require 'model_contract'
```
This should be superfluous.

wielinde replied:
True. Was copy and paste from the user contracts. I fixed it there, too.

ulferts commented on app/contracts/placeholder_users/delete_contract.rb:
```diff
@@ -0,0 +1,58 @@
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+require 'placeholder_users/base_contract'
```
This should be superfluous.

wielinde replied:
True. Was copy and paste from the user contracts. I fixed it there, too.

ulferts commented on app/contracts/placeholder_users/update_contract.rb:
```diff
@@ -0,0 +1,36 @@
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+require 'placeholder_users/base_contract'
```
This ...

wielinde replied:
True. Was copy and paste from the user contracts. I fixed it there, too.

ulferts commented on app/controllers/messages_controller.rb:
```diff
@@ -67,7 +67,7 @@ def new
       .new(user: current_user,
            model: Message.new,
-           contract_class: NoopContract)
+           contract_class: EmptyContract)
```
:+1:

ulferts commented on app/controllers/placeholder_users_controller.rb:
```diff
@@ -0,0 +1,163 @@
+                                               :edit,
+                                               :update,
+                                               :destroy,
+                                               :resend_invitation]
```
`:resend_invitation` does not exist

wielinde replied:
True.

ulferts commented on app/controllers/placeholder_users_controller.rb:
```diff
@@ -0,0 +1,163 @@
+
+  helper_method :gon
+
+  before_action :require_admin, except: [:show, :deletion_info, :destroy]
```
`:deletion_info` does not exist

ulferts replied:
At least for now, the `:destroy` method should also be admin only. This might change in the next PR, though.

wielinde replied:
correct.

ulferts commented on app/contracts/placeholder_users/base_contract.rb:
```diff
@@ -0,0 +1,57 @@
+    attribute :lastname,
+              writeable: ->(*) { user.admin? }
+    attribute :name,
+              writeable: ->(*) { user.admin? }
```
Why is both `lastname` as well as `name` writable? Ideally, from the model layer up, there is only `name`

wielinde replied:
I agree.

wielinde replied:
Hmm, removing `:lastname` here makes other specs fail. In the `PlaceholderUser` model we define `alias_attribute(:name, :lastname)`. Apparently the SetAttributesService detects that also `lastname` gets updated if `name` is changed and reject the change unless we also allow `lastname` to be writable. I will leave it in unless you know of a better solution.

ulferts replied:
You could try using 

`attribute_alias :lastname, :name`

It would still have the DB name in the contract but at least the connection between the two is made visible.

See: https://github.com/opf/openproject/blob/dev/app/contracts/base_contract.rb#L66-L70
And https://github.com/opf/openproject/blob/dev/modules/grids/app/contracts/grids/base_contract.rb#L45

wielinde replied:
When I add `attribute_alias :lastname, :name` I still need to declare the `:lastname` attribute as writable for admins. Or least I don't get it to work otherwise.

ulferts commented on app/controllers/placeholder_users_controller.rb:
```diff
@@ -0,0 +1,163 @@
+  end
+
+  def check_if_deletion_allowed
+    render_404 unless Users::DeleteService.deletion_allowed? @placeholder_user, User.current
```
`Users::DeleteService` is called here instead of `PlaceholderUsers`. Additionally, this would ideally be returned as depending on the result of the DeleteService call.

ulferts replied:
As of now, this results in an error as the `deletion_allowed?` method has been removed from the service.

ulferts commented on app/models/permitted_params.rb:
```diff
@@ -188,6 +188,12 @@ def user
+  def placeholder_user
+    permitted_params = params.require(:placeholder_user).permit(*self.class.permitted_attributes[:placeholder_user])
+
+    permitted_params
```
Why assign it to a variable at all?

wielinde replied:
true.

ulferts commented on app/views/individual_principals/_memberships.html.erb:
```diff
@@ -1,12 +1,12 @@
 <%#-- copyright
 OpenProject is an open source project management software.
-Copyright (C) 2012-2021 the OpenProject GmbH
+Copyright (C) 2012-2020 the OpenProject GmbH
```
This reverts the changes done when updating the copyright.

wielinde replied:
true.

ulferts commented on app/views/individual_principals/_memberships.html.erb:
```diff
@@ -31,14 +31,14 @@ See docs/COPYRIGHT.rdoc for more details.
             <col highlight-col>
             <col highlight-col>
-            <%= call_hook(:view_users_memberships_table_colgroup, user: @user )%>
+            <%= call_hook(:view_placeholder_users_memberships_table_colgroup, placeholder_user: @individual_principal)%>
```
As this is an abstracted view, I'd rather keep the hook name and variable unchanged.

wielinde replied:
And what if we make it even more abstract, such as? Then there would be different hooks called for users and placeholder users. I think that would allow plugins to be more specific.

```
<%= call_hook("view_#{@individual_principal.class.name.underscore.pluralize}_memberships_table_colgroup".to_sym,
                          @individual_principal.class.name.underscore.to_sym => @individual_principal) %>
```

wielinde replied:
extracted to a helper

ulferts commented on app/views/individual_principals/_memberships.html.erb:
```diff
@@ -98,20 +98,20 @@ See docs/COPYRIGHT.rdoc for more details.
                     <% end %>
                   </td>
-                  <%= call_hook(:view_users_memberships_table_row, user: @user, membership: membership, roles: roles, projects: projects )%>
+                  <%= call_hook(:view_placheholder_users_memberships_table_row, placeholder_user: @individual_principal, membership: membership, roles: roles, projects: projects )%>
```
See above

wielinde replied:
see my question above

wielinde replied:
extracted to a helper

ulferts commented on lib/open_project/ui/extensible_tabs.rb:
```diff
@@ -84,6 +85,31 @@ def core_user_tabs
+            #   path: ->(params) { tab_edit_placeholder_user_path(params[:placeholder_user], tab: :groups) },
+            #   label: :label_group_plural,
+            #   if: ->(*) { Group.all.any? }
+            # }
```
Please delete commented out code.

wielinde replied:
Ok.

ulferts commented on spec/contracts/placeholder_users/create_contract_spec.rb:
```diff
@@ -0,0 +1,41 @@
+describe PlaceholderUsers::CreateContract do
+  include_context 'ModelContract shared context'
+
+  let(:placeholder_user) { FactoryBot.create(:placeholder_user) }
```
The model for a `CreateContract` should not be persisted. Ideally, the `PlaceholderUser.new` method is used and attributes (name) assigned to the newly instantiated model.

Every available attribute should then be speced for validations. In this case, this would be that the type be `"PlaceholderUser"` and that the name be present, potentially also the length restriction.

To avoid duplication, part of it might be extracted into a shared examples file to be used by the specs for the CreateContract and UpdateContract.

The specs on contracts need to be among the most thorough of the application as it in part already is and in other will more and more become the central authority for security and data integrity.

wielinde replied:
Done.

ulferts commented on spec/contracts/placeholder_users/update_contract_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  let(:placeholder_user) { FactoryBot.build_stubbed(:placeholder_user) }
+  let(:contract) { described_class.new(placeholder_user, current_user) }
+
+  it_behaves_like 'contract is valid for active admin users only'
```
Please see remarks on `CreateContract`.

wielinde replied:
done

ulferts commented on spec/features/placeholder_users/edit_placeholder_users_spec.rb:
```diff
@@ -0,0 +1,59 @@
+
+  before do
+    allow(User).to receive(:current).and_return current_user
+  end
```
I recently introduced a new wrapper for handling the current user. Instead of:

```
let(:current_user) { admin }

before do
  allow(User).to receive(:current).and_return current_user
end
```

one can now write

```
current_user { admin }
```

wielinde replied:
I could not find that method. Do you mean `login_as admin`? I used that one now.

ulferts commented on spec/features/placeholder_users/index_spec.rb:
```diff
@@ -0,0 +1,74 @@
+require 'spec_helper'
+
+describe 'index placeholder users', type: :feature do
+  let!(:current_user) { FactoryBot.create :admin, created_at: 1.hour.ago }
```
That `created_at` attribute is unnecessary.

One might reuse 
```
using_shared_fixtures :admin
```

ulferts replied:
`using_shared_fixtures` is no more as of #8987

wielinde replied:
True, I removed the created_at attribute.

ulferts reviewed (changes requested):
The general structure is looking good and it is also working as expected, with two tiny exceptions:

* Deleting placeholder users result in an error. This is to be expected as we agreed to exclude it from the PR for now and add it separately. 
* Sorting the placeholder users by name on the index page results in an error.

Please find my code related remarks inline.

wielinde committed:
Add spec for PlaceholderUsersController

wielinde committed:
First code improvements after code review.

- more improvements to come.

wielinde committed:
Further code improvements after review

... still more to come

wielinde committed:
Fix: Make placeholder user contract validate

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/placeholder-user-services

oliverguenther committed:
Correct namespace of delete service

wielinde committed:
Remove :type attribute from base contract of User and PlaceholerUser

...and add it to the CreateContracts.
Also add type validations.

Further extract shared examples for placeholder user attribute
validation

wielinde committed:
Refactor: Extract membership hook calls to helper

wielinde committed:
Fix redirect paths for membership controllers

wielinde committed:
Specs already present in shared exampels.

oliverguenther committed:
Fix duplicates routes for users and placeholder users

oliverguenther committed:
Fix user path

oliverguenther committed:
Add attribute name and lastname

We don't need a writeable check as both are equally writable

oliverguenther committed:
Replace more references to tab_edit_user_path

oliverguenther committed:
Fix module usage of RequireAdminGuard

wielinde committed:
Skip specs for PlaceholderUsers::DeletionService

We will tackle that service in a separate PR.

wielinde committed:
Merge branch 'feature/placeholder-user-services' of github.com:opf/openproject into feature/placeholder-user-services

oliverguenther committed:
Fix group filter for placeholder users

oliverguenther committed:
Fix invalid reference to expect_valid

wielinde committed:
Fix: Fix tabbed edit path for placeholder users

oliverguenther committed:
Fix status filtering on users

oliverguenther committed:
Linting

wielinde committed:
Improve generalisation of individual principal filter cell

- Check for presence of groups and statuses in order
  to toggle visibility of their UI element.
- Remove groups from placeholder user controller and
  cell initialization and options

oliverguenther committed:
Fix selector on groups assign

oliverguenther committed:
Remove using_shared_fixtures

ulferts reviewed (approved):
<img src="https://media.giphy.com/media/kDAqPTNBRfNSxmHLzy/giphy-downsized.gif">
